File: follow car/wechat-miniprogram/utils/ble.js
/**
 * BLE 蓝牙通信工具 — 适配 JDY-33 双模蓝牙模块
 * BLE Service UUID: FFE0, Characteristic UUID: FFE1
 */

const JDY_SERVICE_UUID   = '0000FFE0-0000-1000-8000-00805F9B34FB';
const JDY_CHAR_UUID      = '0000FFE1-0000-1000-8000-00805F9B34FB';
const JDY_NAME_PREFIX    = 'JDY';

/**
 * 通信协议定义
 * 帧格式:  HEAD(0xAA) + CMD(1B) + LEN(1B) + DATA(nB) + CHECKSUM(1B)
 *
 * CMD 列表:
 *   0x01  设置参数   DATA = paramIndex(1B) + value(4B float LE)
 *   0x02  摇杆控制   DATA = x(int8) + y(int8)  范围 -100~+100
 *         x: 左负右正, y: 后负前正, (0,0)=停止
 *   0x03  读取参数   DATA = paramIndex(1B)
 *   0x04  参数应答   DATA = paramIndex(1B) + value(4B float LE)
 */

const CMD = {
  SET_PARAM:  0x01,
  JOYSTICK:   0x02,
  READ_PARAM: 0x03,
  PARAM_ACK:  0x04
};

// 参数索引（与小车端保持一致）
const PARAM_INDEX = {
  FOLLOW_DISTANCE_M:      0,
  OBSTACLE_DIST_CM:       1,
  LIDAR_OBSTACLE_DIST_MM: 2,
  MOTOR_BASE_SPEED:       3,
  MOTOR_TURN_SPEED:       4,
  MOTOR_SLOW_SPEED:       5,
  UWB_TIMEOUT_MS:         6,
  ULTRASONIC_POLL_MS:     7,
  UWB_ANGLE_TOLERANCE_DEG: 8,
  EMERGENCY_STOP_DIST_CM:  9,
  PID_DIST_KP:             10,
  PID_DIST_KI:             11,
  PID_DIST_KD:             12,
  PID_ANGLE_KP:            13,
  PID_ANGLE_KI:            14,
  PID_ANGLE_KD:            15,
  MAX_FOLLOW_SPEED:        16
};

/**
 * 计算校验和 (XOR)
 */
function checksum(bytes) {
  let cs = 0;
  for (let i = 0; i < bytes.length; i++) {
    cs ^= bytes[i];
  }
  return cs & 0xFF;
}

/**
 * 构建发送帧
 */
function buildFrame(cmd, data) {
  const len = data.length;
  const frame = new Uint8Array(3 + len + 1); // HEAD + CMD + LEN + DATA + CS
  frame[0] = 0xAA;
  frame[1] = cmd;
  frame[2] = len;
  for (let i = 0; i < len; i++) {
    frame[3 + i] = data[i];
  }
  frame[3 + len] = checksum(frame.slice(0, 3 + len));
  return frame.buffer;
}

/**
 * 将 float 转为 4 字节 Little-Endian
 */
function floatToBytes(val) {
  const buf = new ArrayBuffer(4);
  new DataView(buf).setFloat32(0, val, true);
  return new Uint8Array(buf);
}

/**
 * 构建设置参数帧
 */
function buildSetParamFrame(paramIndex, value) {
  const valBytes = floatToBytes(value);
  const data = new Uint8Array(5);
  data[0] = paramIndex;
  data.set(valBytes, 1);
  return buildFrame(CMD.SET_PARAM, data);
}

/**
 * 构建摇杆控制帧
 * @param {number} x  左右 -100~+100 (左负右正)
 * @param {number} y  前后 -100~+100 (后负前正)
 */
function buildJoystickFrame(x, y) {
  // 将有符号 int8 转为无符号字节
  const bx = (x < 0) ? (256 + x) : x;
  const by = (y < 0) ? (256 + y) : y;
  return buildFrame(CMD.JOYSTICK, new Uint8Array([bx & 0xFF, by & 0xFF]));
}

/**
 * 解析接收帧
 */
function parseFrame(buffer) {
  const bytes = new Uint8Array(buffer);
  if (bytes.length < 4 || bytes[0] !== 0xAA) return null;
  const cmd = bytes[1];
  const len = bytes[2];
  if (bytes.length !== 3 + len + 1) return null;
  const cs = checksum(bytes.slice(0, 3 + len));
  if (cs !== bytes[3 + len]) return null;
  const data = bytes.slice(3, 3 + len);
  return { cmd, data };
}

module.exports = {
  JDY_SERVICE_UUID,
  JDY_CHAR_UUID,
  JDY_NAME_PREFIX,
  CMD,
  PARAM_INDEX,
  buildSetParamFrame,
  buildJoystickFrame,
  parseFrame,
  floatToBytes
};

File: follow car/stm32_follow_car/Src/ble.c
/*
 * ble.c - JDY-33 蓝牙通信协议 (USART1)
 *
 * 接收微信小程序的参数修改和摇杆控制指令
 * 帧格式: 0xAA + CMD + LEN + DATA + XOR校验
 */
#include "ble.h"
#include "motor.h"

/* ========== 全局变量 ========== */
BLE_RxCtx_t              g_ble_rx;
volatile int8_t          g_ble_joy_x       = 0;
volatile int8_t          g_ble_joy_y       = 0;
volatile bool            g_ble_manual_active = false;
volatile uint32_t        g_ble_joy_tick    = 0;

/* ========== 调试变量 (Keil Watch窗口监控) ========== */
volatile uint32_t dbg_irq_cnt     = 0;  /* 中断触发总次数 */
volatile uint8_t  dbg_last_byte   = 0;  /* 最后收到的原始字节 */
volatile uint8_t  dbg_state       = 0;  /* 当前状态机状态 */
volatile uint8_t  dbg_frame_cmd   = 0;  /* 最后一帧的CMD */
volatile uint8_t  dbg_frame_len   = 0;  /* 最后一帧的LEN */
volatile uint8_t  dbg_cs_calc     = 0;  /* 计算的校验值 */
volatile uint8_t  dbg_cs_recv     = 0;  /* 收到的校验值 */
volatile uint32_t dbg_frame_ok    = 0;  /* 校验通过的帧计数 */
volatile uint32_t dbg_frame_fail  = 0;  /* 校验失败的帧计数 */
volatile int8_t   dbg_joy_x       = 0;  /* 最后解析的摇杆X */
volatile int8_t   dbg_joy_y       = 0;  /* 最后解析的摇杆Y */

/* ========== 内部函数 ========== */

/* 发送一个字节到 USART1 */
static void usart1_send_byte(uint8_t byte)
{
    while (!(huart1.Instance->SR & USART_SR_TXE)) {}
    huart1.Instance->DR = byte;
}

/* 发送一帧 */
static void ble_send_frame(uint8_t cmd, const uint8_t *data, uint8_t len)
{
    uint8_t cs = BLE_FRAME_HEAD ^ cmd ^ len;
    usart1_send_byte(BLE_FRAME_HEAD);
    usart1_send_byte(cmd);
    usart1_send_byte(len);
    for (uint8_t i = 0; i < len; i++) {
        usart1_send_byte(data[i]);
        cs ^= data[i];
    }
    usart1_send_byte(cs);
}

/* 根据索引设置参数 */
static void set_param(uint8_t index, float value)
{
    switch (index) {
    case 0:  g_car_params.follow_distance_m = value; break;
    case 1:  g_car_params.obstacle_dist_cm = value; break;
    case 2:  g_car_params.lidar_obstacle_dist_mm = value; break;
    case 3:  g_car_params.motor_base_speed = value; break;
    case 4:  g_car_params.motor_turn_speed = value; break;
    case 5:  g_car_params.motor_slow_speed = value; break;
    case 6:  g_car_params.uwb_timeout_ms = value; break;
    case 7:  g_car_params.ultrasonic_poll_ms = value; break;
    case 8:  g_car_params.uwb_angle_tolerance_deg = value; break;
    case 9:  g_car_params.emergency_stop_dist_cm = value; break;
    case 10: g_car_params.pid_dist_kp = value; break;
    case 11: g_car_params.pid_dist_ki = value; break;
    case 12: g_car_params.pid_dist_kd = value; break;
    case 13: g_car_params.pid_angle_kp = value; break;
    case 14: g_car_params.pid_angle_ki = value; break;
    case 15: g_car_params.pid_angle_kd = value; break;
    case 16: g_car_params.max_follow_speed = value; break;
    default: break;
    }
}

/* 根据索引读取参数 */
static float get_param(uint8_t index)
{
    switch (index) {
    case 0:  return g_car_params.follow_distance_m;
    case 1:  return g_car_params.obstacle_dist_cm;
    case 2:  return g_car_params.lidar_obstacle_dist_mm;
    case 3:  return g_car_params.motor_base_speed;
    case 4:  return g_car_params.motor_turn_speed;
    case 5:  return g_car_params.motor_slow_speed;
    case 6:  return g_car_params.uwb_timeout_ms;
    case 7:  return g_car_params.ultrasonic_poll_ms;
    case 8:  return g_car_params.uwb_angle_tolerance_deg;
    case 9:  return g_car_params.emergency_stop_dist_cm;
    case 10: return g_car_params.pid_dist_kp;
    case 11: return g_car_params.pid_dist_ki;
    case 12: return g_car_params.pid_dist_kd;
    case 13: return g_car_params.pid_angle_kp;
    case 14: return g_car_params.pid_angle_ki;
    case 15: return g_car_params.pid_angle_kd;
    case 16: return g_car_params.max_follow_speed;
    default: return 0.0f;
    }
}

/* ========== 公开接口 ========== */

void BLE_Init(void)
{
    memset(&g_ble_rx, 0, sizeof(g_ble_rx));
    g_ble_rx.state = BLE_RX_WAIT_HEAD;
    g_ble_joy_x = 0;
    g_ble_joy_y = 0;
    g_ble_manual_active = false;
    g_ble_joy_tick = 0;
}

void BLE_SendParamAck(uint8_t paramIndex, float value)
{
    uint8_t data[5];
    data[0] = paramIndex;
    memcpy(&data[1], &value, 4);
    ble_send_frame(BLE_CMD_PARAM_ACK, data, 5);
}

/* USART1 接收中断 — 逐字节状态机 */
void BLE_UART_IRQHandler(void)
{
    if (__HAL_UART_GET_FLAG(&huart1, UART_FLAG_RXNE)) {
        uint8_t byte = (uint8_t)(huart1.Instance->DR & 0xFF);

        dbg_irq_cnt++;
        dbg_last_byte = byte;
        dbg_state = (uint8_t)g_ble_rx.state;

        /* 如果上一帧还没处理完，丢弃新数据等待处理 */
        if (g_ble_rx.ready) return;

        switch (g_ble_rx.state) {
        case BLE_RX_WAIT_HEAD:
            if (byte == BLE_FRAME_HEAD) {
                g_ble_rx.state = BLE_RX_WAIT_CMD;
            }
            break;

        case BLE_RX_WAIT_CMD:
            g_ble_rx.cmd = byte;
            dbg_frame_cmd = byte;
            g_ble_rx.state = BLE_RX_WAIT_LEN;
            break;

        case BLE_RX_WAIT_LEN:
            g_ble_rx.len = byte;
            g_ble_rx.idx = 0;
            dbg_frame_len = byte;
            if (byte == 0 || byte > BLE_RX_BUF_SIZE) {
                g_ble_rx.state = BLE_RX_WAIT_HEAD; /* 非法长度 */
            } else {
                g_ble_rx.state = BLE_RX_WAIT_DATA;
            }
            break;

        case BLE_RX_WAIT_DATA:
            g_ble_rx.buf[g_ble_rx.idx++] = byte;
            if (g_ble_rx.idx >= g_ble_rx.len) {
                g_ble_rx.state = BLE_RX_WAIT_CS;
            }
            break;

        case BLE_RX_WAIT_CS: {
            /* XOR 校验 */
            uint8_t cs = BLE_FRAME_HEAD ^ g_ble_rx.cmd ^ g_ble_rx.len;
            for (uint8_t i = 0; i < g_ble_rx.len; i++) {
                cs ^= g_ble_rx.buf[i];
            }
            dbg_cs_calc = cs;
            dbg_cs_recv = byte;
            if (cs == byte) {
                g_ble_rx.ready = 1;
                dbg_frame_ok++;
            } else {
                dbg_frame_fail++;
            }
            g_ble_rx.state = BLE_RX_WAIT_HEAD;
            break;
        }
        }
    }

    /* 清除溢出错误，防止卡死 */
    if (__HAL_UART_GET_FLAG(&huart1, UART_FLAG_ORE)) {
        __HAL_UART_CLEAR_OREFLAG(&huart1);
    }
}

/* 主循环调用 — 处理完整帧 */
void BLE_Poll(void)
{
    if (!g_ble_rx.ready) {
        /* 手动控制超时检查 */
        if (g_ble_manual_active &&
            (HAL_GetTick() - g_ble_joy_tick > BLE_MANUAL_TIMEOUT_MS)) {
            g_ble_joy_x = 0;
            g_ble_joy_y = 0;
            g_ble_manual_active = false;
        }
        return;
    }

    /* 处理帧 */
    switch (g_ble_rx.cmd) {
    case BLE_CMD_SET_PARAM:
        if (g_ble_rx.len >= 5) {
            uint8_t idx = g_ble_rx.buf[0];
            float val;
            memcpy(&val, &g_ble_rx.buf[1], 4);
            set_param(idx, val);
            /* 回传确认 */
            BLE_SendParamAck(idx, val);
        }
        break;

    case BLE_CMD_JOYSTICK:
        if (g_ble_rx.len >= 2) {
            g_ble_joy_x = (int8_t)g_ble_rx.buf[0];
            g_ble_joy_y = (int8_t)g_ble_rx.buf[1];
            dbg_joy_x = g_ble_joy_x;
            dbg_joy_y = g_ble_joy_y;
            /* 限幅 */
            if (g_ble_joy_x > 100)  g_ble_joy_x = 100;
            if (g_ble_joy_x < -100) g_ble_joy_x = -100;
            if (g_ble_joy_y > 100)  g_ble_joy_y = 100;
            if (g_ble_joy_y < -100) g_ble_joy_y = -100;
            g_ble_joy_tick = HAL_GetTick();
            g_ble_manual_active = (g_ble_joy_x != 0 || g_ble_joy_y != 0);
        }
        break;

    case BLE_CMD_READ_PARAM:
        if (g_ble_rx.len >= 1) {
            uint8_t idx = g_ble_rx.buf[0];
            float val = get_param(idx);
            BLE_SendParamAck(idx, val);
        }
        break;

    default:
        break;
    }

    g_ble_rx.ready = 0;
}
